third_party/2and3/click/types.pyi

```diff
@@ -55,11 +65,22 @@ class IntParamType(ParamType):
     def convert(self, value: str, param: Optional[Parameter], ctx: Optional[Context]) -> int: ...
 
 class IntRange(IntParamType):
+    min: Optional[int]
+    max: Optional[int]
+    clamp: bool
     def __init__(self, min: Optional[int] = ..., max: Optional[int] = ..., clamp: bool = ...) -> None: ...
 
 _PathType = TypeVar("_PathType", str, bytes)
 
 class Path(ParamType):
+    exists: bool
+    file_okay: bool
+    dir_okay: bool
+    writable: bool
+    readable: bool
+    resolve_path: bool
+    allow_dash: bool
+    type: Optional[Type[_PathType]]
```

Comment:
Not sure on the typing for this causes the tests to fail.
I already tried fixing it according to the hints produced by the failing test, but had not luck.
Please advise

Reply:
The typevar doesn't make sense in this context, since there's nothing to bind it. Perhaps you want `Optional[Union[Type[str], Type[bytes]]]`.

Reply:
I see it seems I slightly misunderstood TypeVars and binding until now. I've added the type as per your suggestion since I also believe this to be the correct hint for this. For this I added it as an alias `_PathTypeBound` just below the `_PathType` TypeVar. As to prevent possible future type changes only being done to the `_PathType` init argument and not the attribute, especially since they argument and attribute do not share the exact same name.

Also rebased to the latest master containing the proposed fix for #4815 .

Reply:
The typevar is actually already broken in the current version: It's used in two different functions without the class being generic. I see two possible solutions:

1. Make the class generic over the path type and use the typevar in all three positions.
2. Use the union in all three positions.

I have a slight preference for the first option, although with the usual caveat that this requires every user of `Path` to annotate it. (At least as long as we don't have support for generic defaults.) Cc @JelleZijlstra.

Reply:
Making the class generic seems to make sense; in that case we should add an overload to the `__init__` function to specify the value of the typevar when the path argument isn't passed.

Reply:
@max-frank Do you want to change it? Otherwise I would merge the PR as is and send a separate PR myself.

Reply:
@srittau Ah yes sorry I was doing a deep dive into the click code base and got a bit side tracked.
If you wan't you can do a separate PR I was just in the process of figuring out what the default would actually be when `path_type=None`. 

From what I can see it should boil down to the following two cases depending on the system
- `unicode`: for `PY2 and WIN` (read as boolean expression)
- `str`: for all other cases

One question on the  `__init__` overload would that look like the following:
```python
_PathTypeDefault = unicode if PY2 and WIN else str
...
def __init__(self: Path[_PathTypeDefault], ..., path_type: Literal[None] = None):
         ...
```
------
The above conclusion is derived from the following click code:

Special argv parsing for `PY2 and WIN`, other cases use normal `sys.argv` which gives us `List[str]`

https://github.com/pallets/click/blob/1784558ed7c75c65764d2a434bd9cbb206ca939d/src/click/utils.py#L356
```python
if PY2 and WIN and _initial_argv_hash == _hash_py_argv():
        return _get_windows_argv()
    return sys.argv[1:]
```
with the windows function using windows libs and some custom code to get `List[unicode]`.
https://github.com/pallets/click/blob/1784558ed7c75c65764d2a434bd9cbb206ca939d/src/click/_winconsole.py#L307

For Path options/args eventually one list element will make it to the `Path.convert` function which in turn calls 
[coerce_path_result(self, rv)](https://github.com/pallets/click/blob/1784558ed7c75c65764d2a434bd9cbb206ca939d/src/click/types.py#L590) with the value as rv
```python
if self.type is not None and not isinstance(rv, self.type):
            if self.type is text_type:
                rv = rv.decode(get_filesystem_encoding())
            else:
                rv = rv.encode(get_filesystem_encoding())
        return rv
```

Now the above is for the 7.1.2 code base 8.0 drops support for python2.7 so the custom will be removed (i.e., for the pre-release it is already).

Reply:
@srittau Another side node related the above in the click code review process I noticed that the type hints on the Path functions are not entirely correct.

```python
def coerce_path_result(self, rv: Union[str, bytes]) -> _PathType: ...
def __call__(self, value: Optional[str], param: Optional[Parameter] = ..., ctx: Optional[Context] = ...) -> _PathType: ...
def convert(self, value: str, param: Optional[Parameter], ctx: Optional[Context]) -> _PathType: ...
```
I believe the `value` arg on `__call__`  and `convert` should match `rv` from `coerce_path_result`.